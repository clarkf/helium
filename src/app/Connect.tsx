import { useState } from "react";

import { FILTER_LAZY_LOAD, discover, login, sync } from "../matrix";
import { useApp } from "../state";

interface Props {
  onConnect?: () => void;
}

export default function Connect({ onConnect }: Props): JSX.Element {
  const handleSync = useApp((s) => s.handleSync);
  const [host, setHost] = useState("");
  const [userId, setUserId] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState<string | null>(null);

  function handleSubmit(e: React.FormEvent): void {
    e.preventDefault();
    connect(host, userId, password)
      .then(({ discovery, token }) => {
        sync(discovery, token, { filter: FILTER_LAZY_LOAD }).then(handleSync);
        onConnect?.();
      })
      .catch((e) => setError(e.toString()));
  }

  return (
    <form onSubmit={handleSubmit}>
      {error ? <div role="alert">{error}</div> : null}
      <div>
        <label htmlFor="connect-host">Host</label>
        <input
          id="connect-host"
          type="text"
          value={host}
          onChange={(e) => setHost(e.currentTarget.value)}
        />
      </div>

      <div>
        <label htmlFor="connect-user">User ID</label>
        <input
          id="connect-user"
          type="text"
          value={userId}
          onChange={(e) => setUserId(e.currentTarget.value)}
        />
      </div>

      <div>
        <label htmlFor="connect-password">Password</label>
        <input
          id="connect-password"
          type="password"
          value={password}
          onChange={(e) => setPassword(e.currentTarget.value)}
        />
      </div>

      <div>
        <button type="submit">Connect</button>
      </div>
    </form>
  );
}

async function connect(host: string, user: string, password: string) {
  const discovery = await discover(host);
  const token = await login(discovery, user, password);

  console.debug({ discovery, token });

  return { discovery, token };
}
