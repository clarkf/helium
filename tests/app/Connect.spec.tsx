// @vitest-environment happy-dom

import {
  describe,
  it,
  vi,
  expect,
  beforeAll,
  afterAll,
  afterEach,
} from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { setupServer } from "msw/node";
import { rest } from "msw";

import Connect from "../../src/app/Connect";
import { resolveDiscovery, resolveEmptySync, resolveLogin } from "../handlers";

describe("<Connect />", () => {
  const server = setupServer(
    rest.get(
      "https://example.com/.well-known/matrix/client",
      resolveDiscovery("https://matrix.example.com"),
    ),
    rest.get(
      "https://404.example.com/.well-known/matrix/client",
      (_, res, ctx) => res(ctx.status(404), ctx.xml("<b>Not found!</b>")),
    ),
    rest.post(
      "https://matrix.example.com/_matrix/client/v3/login",
      resolveLogin((user, pass) => user == "user" && pass == "p@ssword"),
    ),
    rest.get(
      "https://matrix.example.com/_matrix/client/v3/sync",
      resolveEmptySync,
    ),
  );
  beforeAll(() => server.listen({ onUnhandledRequest: "error" }));
  afterAll(() => server.close());
  afterEach(() => server.resetHandlers());

  async function fillOut(host: string, userId: string, pass: string) {
    const user = userEvent.setup();
    await user.type(screen.getByLabelText("Host"), host);
    await user.type(screen.getByLabelText("User ID"), userId);
    await user.type(screen.getByLabelText("Password"), pass);
    await user.click(screen.getByRole("button"));
  }

  it("renders", () => {
    const { asFragment } = render(<Connect />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("discovers the server", async () => {
    const spy = vi.fn();
    render(<Connect onConnect={spy} />);

    await fillOut("example.com", "user", "p@ssword");

    expect(spy).toHaveBeenCalledOnce();
  });

  it("displays an error on 404", async () => {
    render(<Connect />);

    await fillOut("404.example.com", "cool-user", "p@ssword");

    const err = await screen.findByRole("alert");
    expect(err).toBeVisible();
  });

  it("displays an error on invalid credentials", async () => {
    render(<Connect />);

    await fillOut("example.com", "bad-user", "pXssword");

    const err = await screen.findByRole("alert");
    expect(err).toBeVisible();
  });
});
