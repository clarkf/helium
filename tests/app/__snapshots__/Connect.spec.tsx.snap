// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Connect /> > renders 1`] = `
<DocumentFragment>
  <form>
    <div>
      <label
        for="connect-host"
      >
        Host
      </label>
      <input
        id="connect-host"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="connect-user"
      >
        User ID
      </label>
      <input
        id="connect-user"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="connect-password"
      >
        Password
      </label>
      <input
        id="connect-password"
        type="password"
        value=""
      />
    </div>
    <div>
      <button
        type="submit"
      >
        Connect
      </button>
    </div>
  </form>
</DocumentFragment>
`;
